Assets/Scripts/Interaction/PaintableSurface.cs

```diff
@@ -0,0 +1,89 @@
+    public class PaintableSurface : MonoBehaviour, IContinuousInteraction
+    {
+        [SerializeField] private InteractionKind kind = null;
+        
+        public void Setup(Action<InteractionKind> listen, Action<InteractionKind> blocks)
+        {
+            listen(kind);
+        }
+
+        public IEnumerator Handle(InteractionEvent e)
+        {
+            if (e.sender == null)
+            {
+                Debug.LogWarning($"Tried to paint null on ({name})");
+                yield break;
+            }
+
+            Painter painter = e.sender.GetComponent<Painter>();
+            if (painter == null)
+            {
+                Debug.LogError($"Tried to paint with {e.sender} without a Painter component.)");
+                yield break;
+            }
+
+            PlayerInteractor playerInteractor = e.player.GetComponent<PlayerInteractor>();
+            if (playerInteractor == null)
+            {
+                Debug.LogError($"Received event from player {e.player} without a PlayerInteractor component.");
+                yield break;
+            }
+            
+            var playerHands = e.player.GetComponent<Hands>();
+            if (playerHands == null)
+            {
+                Debug.LogWarning($"Player painting with object ({e.sender.name}) does not have a hands component");
+                yield break;
+            }
+            
+            NetworkInteractionSpawner networkInteractionSpawner = e.player.GetComponent<NetworkInteractionSpawner>();
+            if (networkInteractionSpawner == null)
+            {
+                Debug.LogError($"Received paint event from player {e.player} without a NetworkInteractionSpawner component.");
+                yield break;
+            }
+
+            var inventoryUi = FindObjectOfType<UIInventory>();
+
+            while (e.runWhile.Invoke(e))
+            {
+                float remainingSupply = painter.GetRemainingSupplyPercentage();
+                
+                //Update paint location and stop painting if surface no longer applicable
+                if (!playerInteractor.GetWorldData(out var hit)) yield return null;
+                if (hit.transform.GetComponent<PaintableSurface>() == null) yield break;
+                //Update remaining supply bar on painter
+                if(inventoryUi) inventoryUi.SetHeldItemSupply(remainingSupply);
+
+                //Decrease supply and spawn new decal. Set to be slightly above collider to avoid clipping.
+                playerInteractor.CmdDecreaseSupplyOfItem(painter.gameObject);
+                Vector3 position = hit.point + Vector3.up * 0.01f;
+                Quaternion rotation = Quaternion.FromToRotation(Vector3.back, hit.normal);
+                networkInteractionSpawner.CmdSpawnPainterDecal(painter.Properties.name, position, rotation);
+
+                if (remainingSupply <= 0)
+                {
+                    playerHands.DestroyItemInHand();
```

Comment:
This might just be a note for when we properly do the inventory, but I think objects should be able to destroy themselves without needed a reference to the hand.

Reply:
I agree. I had to do it this way to make it work with the current inventory system. If we come up with a better approach, I'd absolutely redo it.